### PR TITLE
Require permitted scopes when registering a client

### PR DESCRIPTION
**Pull request by angusmcleod**
@pmusaraj After working with the ActivityPub implementation of user api key client registration I think it makes sense to restrict external client registration to clients restricted by permitted scopes. 

This means that you can only register a client if you also define the scopes of the keys that can be created with the client, and those scopes are allowed by the site owner. This still allows any client (i.e. any domain) to be registered to allow for the one-to-many use case (e.g. ActivityPub), but it improves the security and control for the site owner.

In the case of the ActivityPub plugin the way it will work is as follows:

- ActivityPub plugin will define a new user api key scope that permits access to `/ap/local/actor/find-by-user` (currently the only protected route we need to access)
- Site owner will need to add ActivityPub scope to both `allow_user_api_key_scopes` and `allow_user_api_key_client_scopes` if they wish to allow any external site to register a client.
- Any external site can register a client with that scope and initiate the user api key flow, allowing them to _only_ generate user api keys with the ActivityPub scope.

Existing user api key flow continues to work as normal. Clients created automatically in the UserApiKeysController as part of the normal flow will not be created with scopes (key scopes (as opposed to client scopes) will continue to work as normal).

### CODE REVIEW

File: app/controllers/user_api_key_clients_controller.rb [hunk 1]
```diff
@@ -2,26 +2,36 @@
 class UserApiKeyClientsController < ApplicationController
   layout "no_ember"
 
-  requires_login
-  skip_before_action :check_xhr, :preload_json
+  skip_before_action :check_xhr, :preload_json, :verify_authenticity_token
```
Comment (pmusaraj) [inline app/controllers/user_api_key_clients_controller.rb:5]: Is it possible at all to keep this limited to authenticated requests? I get why we may wish to open this up to anons, the client row may need to be created by an unauthenticated third party. I'm a bit concerned though that this opens up a potential attack vector, a carefully crafted request could create dummy clients. It's mitigated by scopes to some extent...

I guess a malicious actor could also create a user and exploit this that way as well. But at least there would be a bit more tracing.

**Review by pmusaraj — COMMENTED** [review]
Looks good overall, thanks!

File: app/controllers/user_api_key_clients_controller.rb (hunk 1 above)
Comment (angusmcleod) [reply app/controllers/user_api_key_clients_controller.rb:5]: I did a fair bit of thinking on this this week, and came up with the following: https://github.com/angusmcleod/discourse/commit/a87176a39b3547162fc012d7aa4c5b056a4815f0

Adds:
- a relatively strict rate-limit for the "create" action.
- Only permits creation, not updates. This ensures that the public key and auth_redirect associated with a client always remains the same. If you want to change these for a "registered" client you have to create a new one with a different id.
- A cleanup job to handle "stale" clients (where all the client's keys are older than x days).
- Make the `allow_user_api_key_client_scopes` site setting hidden, to reduce the chance of an admin adding overly broad scopes to this allow-list. Moreover this functionality, i.e. 1:many will likely always be managed programmatically anyway.

Does that ally your concerns here?

**Comment by angusmcleod** [comment]
In addition to my note above, I've also committed the WIP implementation of this in the ActivityPub plugin here so you can see this working in practice: https://github.com/discourse/discourse-activity-pub/pull/133

File: app/controllers/user_api_key_clients_controller.rb (hunk 1 above)
Comment (pmusaraj) [reply app/controllers/user_api_key_clients_controller.rb:5]: Yes, thank you! That does indeed address my concerns.

**Review by pmusaraj — APPROVED** [review]
This looks good, approving, will merge Monday as I start my week. Thanks!

**Pull request merged**